File: src/dots/core/template.py
"""Template engine for dependency URL resolution."""

import platform
from typing import Optional


def get_system_arch() -> str:
    """Detect system architecture (x86_64, aarch64, etc)."""
    machine = platform.machine().lower()
    if machine in ["x86_64", "amd64"]:
        return "x86_64"
    elif machine in ["aarch64", "arm64"]:
        return "aarch64"
    return machine


def resolve_arch(arch_map: Optional[dict[str, str]]) -> str:
    """
    Resuelve la arquitectura final.

    Si hay arch_map, traduce. Si no, devuelve el valor bruto del sistema.
    """
    raw = get_system_arch()
    if arch_map:
        return arch_map.get(raw, raw)
    return raw


def render(template: str, context: dict[str, str]) -> str:
    """
    Reemplaza {{key}} con context[key] en el template.

    >>> render("fd-{{version}}-{{arch}}.tar.gz", {"version": "v8.7.0", "arch": "x86_64"})
    'fd-v8.7.0-x86_64.tar.gz'
    """
    result = template
    for key, value in context.items():
        placeholder = f"{{{{{key}}}}}"
        result = result.replace(placeholder, value)
    return result


def build_context(version: Optional[str], arch_map: Optional[dict]) -> dict[str, str]:
    """
    Construye el contexto de templating para una dependency.
    """
    return {
        "arch": resolve_arch(arch_map),
        "version": version or "",
    }

File: internal/template/template.go
package template

import (
	"regexp"
	"runtime"
	"strings"
)

var placeholderRe = regexp.MustCompile(`\{\{[^}]+}}`)

// GetSystemArch detects the system architecture.
// Returns "x86_64", "aarch64", or the raw machine value.
func GetSystemArch() string {
	switch runtime.GOARCH {
	case "amd64":
		return "x86_64"
	case "arm64":
		return "aarch64"
	default:
		return runtime.GOARCH
	}
}

// ResolveArch translates architecture using an arch map if provided.
func ResolveArch(archMap map[string]string) string {
	raw := GetSystemArch()
	if archMap != nil {
		if mapped, ok := archMap[raw]; ok {
			return mapped
		}
	}
	return raw
}

// Render replaces {{key}} placeholders with values from the context map.
// Any remaining unresolved placeholders (missing keys) are removed.
func Render(tmpl string, context map[string]string) string {
	result := tmpl
	for key, value := range context {
		placeholder := "{{" + key + "}}"
		result = strings.ReplaceAll(result, placeholder, value)
	}
	// Remove any remaining unresolved placeholders
	result = placeholderRe.ReplaceAllString(result, "")
	return result
}

// BuildContext builds a template context with arch and optional version.
func BuildContext(version string, archMap map[string]string) map[string]string {
	return map[string]string{
		"arch":    ResolveArch(archMap),
		"version": version,
	}
}
